### DISCUSSION TITLE
Restrict entry types to certain levels in Structure sections

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8118] [Ideas]
> Created by: Brandon Kelly (brandon@pixelandtonic.com) on 2015/03/06 03:54:45 +0000
> Votes at time of UserVoice import: 32

-----------------------------------

It would be nice if it was possible to restrict Structure sections' entry types to certain levels.

For example, on a section like http://buildwithcraft.com/docs, where there is are "Heading" and "Documentation Page" entry types, it would be nice if there was a way to restrict the top level to just Heading entries, and all other levels to just Documentation Page entries.

**Comment by angrybrad** [comment]
> Posted by Thijs Kaspers (thijs@24creative.nl) on 2015/03/11 08:43:15 +0000
-----------------------------------

I agree! It feels weird that all structure levels use the same entry type by default, because in most cases they will need different fields...

Either add an option to set a default entrytype for each level, or the option to restrict entry types to certain levels. Or a combination of both ;)

**Comment by jordanlev** [comment]
This is a much-needed feature for my company as well. Here's an alternate solution that I think is conceptually simple and offers a lot of flexibility (taken from how ProcessWire handles things): If you could set an entry type to only allow certain other entry types underneath it (or vice-versa: entry types could be set to only be allowed underneath other entry type(s)), then you could do all sorts of things...

For example, if I have an employee directory and my entry types are "Employee" and "Department", I want each employee to be in 1 and only 1 department. I could set the section to only allow "Department" entry types at the top-level, then set the "Department" entry type to only allow "Employee" entries as children. (Or maybe it's vice-versa... saying that an entry type can only be a child of some other entry type, so a "Department" cannot be a child of any other type, and "Employee" can only be a child of "Department" entry type).

But if I have a more complex structure (to take an example from a real project I'm working on): I've got product categories and products. Categories can be nested, but products must be under 1 category (and a category either has sub-categories or sub-products, but not both). So I could have 3 entry types in this section: "category that has subcategories", "category that has products", and "product", then set each of them to only allow the children of the proper type (and set "product" to not allow child entries at all). In this way I can have an arbitrarily-nested structure (I don't have to set it to a specific number of levels) but also enforce the hierarchy I want.

**Comment by brandonkelly** [comment]
@jordanlev Thanks for the feedback! Setting allowed entry types based on parent/child relationships sounds much nicer than restricting per level. Could you post this as a new "issue" though? It’s more likely we’d implement your idea, and it would be nice to link directly to it in the changelog.

**Comment by jordanlev** [comment]
@brandonkelly done, thanks for listening :)
I know you can't commit to any dates, but as a general question... are all new features going into Craft 3 at this point, or will you still be adding new features to the 2.x line while 3 is in beta?

**Comment by brandonkelly** [comment]
At this point we’re only adding new features to v2 if it will be easy to port to v3. Considering how much has changed under the hood in Craft 3, that means the vast majority of new features are exclusive to v3, but not all.

**Comment by MaartenVenema** [comment]
I know this is quite an old request but i would like to breath some new life into it. There are quite a few occasions where this would really come in handy. For example: I have a product and i want to have variants of that product. Products can only be on level 1 and variant should only go in level 2.

**Comment by brandonkelly** [comment]
In the meantime, check out the [Level Entry Types](https://plugins.craftcms.com/level-entry-types) plugin, which provides this functionality.

**Comment by MaartenVenema** [comment]
Thanks for mentioning that plugin @brandonkelly. Will have a look. Would still be really cool to see this in CraftCMS itself. Thanks for responding either way 👍

**Comment by jurriaanr** [comment]
And here we are. 3 years later and this is still not implemented? Seems such a logical feature to me? either by level or hierarchy but you should be able to limit where the entry type is placed!? The mentioned levels plugin is no longer maintained and to be frank in a type called structure this should not be a plugin but default behaviour

**Reply by brandonkelly** [reply]
We are hoping to make this possible in Craft 5 via conditions on entry types.

FWIW, we get several new feature requests every day. There’s no way we could ever get to all of them (nor should we). We do our best to prioritize around popularity and how well features fit into the overall product planning & strategy.

**Reply by jurriaanr** [reply]
Fair enough and I know you guys are working hard and are really busy, I was just sort of surprised that this wasnt a higher priority not so much for you guys but from users

**Comment by michaelrog** [comment]
This is straightforward to customize using `Entry::EVENT_DEFINE_ENTRY_TYPES`.

(Actually, I imagine building first-party UI around this would be _muuuuch_ more complicated: It's hard to envision a one-size-fits-all solution for this, whereas it's easy to add a few conditional lines to my module to directly accomplish whatever project-specific customization I need.)
